arrow-libs/fx/arrow-collectors/src/commonMain/kotlin/arrow/collectors/Collect.kt

```diff
@@ -0,0 +1,150 @@
+ * @param concurrency Defines the concurrency limit
+ */
+@OptIn(FlowPreview::class)
+public suspend fun <T, R> Iterable<T>.parCollect(
+  collector: Collector<T, R>,
+  concurrency: Int = DEFAULT_CONCURRENCY,
+): R = asFlow().collect(collector, concurrency)
+
+/**
+ * Performs collection over the elements of [this].
+ * The amount of concurrency depends on the
+ * [Characteristics] of [collector], and can
+ * be tweaked using the [concurrency] parameter.
+ *
+ * [this] is iterated only once during collection.
+ *
+ * Note: if you need to perform changes on the values
+ * before collection, we strongly recommend to convert
+ * the [Sequence] into a [Flow] using [asFlow],
+ * perform those changes, and then using [collect].
+ *
+ * @receiver Sequence of values to collect
+ * @param collector Describes how to collect the values
+ * @param concurrency Defines the concurrency limit
+ */
+@OptIn(FlowPreview::class)
+public suspend fun <T, R> Sequence<T>.parCollect(
+  collector: Collector<T, R>,
+  concurrency: Int = DEFAULT_CONCURRENCY,
+): R = asFlow().collect(collector, concurrency)
+
+/**
+ * Performs collection over the elements of [this]
+ * in a non-concurrent fashion.
+ *
+ * [this] is iterated only once during collection.
+ *
+ * @receiver Sequence of values to collect
+ * @param collector Describes how to collect the values
+ */
+public fun <T, R> Iterable<T>.collect(
+  collector: NonSuspendCollector<T, R>
+): R = iterator().collectI(collector)
+
+/**
+ * Performs collection over the elements of [this]
+ * in a non-concurrent fashion.
+ *
+ * [this] is iterated only once during collection.
+ *
+ * @receiver Sequence of values to collect
+ * @param collector Describes how to collect the values
+ */
+public fun <T, R> Sequence<T>.collect(
+  collector: NonSuspendCollector<T, R>
+): R = iterator().collectI(collector)
+
+public fun <T, R> Iterator<T>.collect(
+  collector: NonSuspendCollector<T, R>
+): R = collectI(collector)
+
+@OptIn(FlowPreview::class, ExperimentalCoroutinesApi::class)
+@Suppress("UNINITIALIZED_VARIABLE", "UNCHECKED_CAST")
+internal suspend fun <A, T, R> Flow<T>.collectI(
+  collector: CollectorI<A, T, R>,
+  concurrency: Int = DEFAULT_CONCURRENCY,
+): R {
+  var accumulator: A
+  val started = this.onStart { accumulator = collector.supply() }
+  val continued = when {
+    Characteristics.CONCURRENT in collector.characteristics -> when {
+      Characteristics.UNORDERED in collector.characteristics ->
+        started.parMapUnordered(concurrency) { collector.accumulate(accumulator, it) }
+
+      else ->
+        started.parMap(concurrency) { collector.accumulate(accumulator, it) }
+    }
+
+    else -> started.map { collector.accumulate(accumulator, it) }
+  }
+  var completed: R
+  continued.onCompletion {
+    completed = when {
+      Characteristics.IDENTITY_FINISH in collector.characteristics -> accumulator as R
+      else -> collector.finish(accumulator)
+    }
+  }.collect { }
+  return completed
```

Comment:
Hi, I think we can simplify this by `collect()` then returning.
It has the same the behavior.

```suggestion
  continued.collect()

  return when {
    Characteristics.IDENTITY_FINISH in collector.characteristics -> accumulator as R
    else -> collector.finish(accumulator)
  }
```